sdk/core/Azure.Core.Experimental/src/Auth/OAuth2BearerTokenAuthenticationPolicy.cs

```diff
@@ -0,0 +1,93 @@
+    {
+        _tokenProvider = tokenProvider;
+        _flowContext = GetContext(contexts, tokenProvider);
+    }
+
+    /// <inheritdoc />
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        ProcessAsync(message, pipeline, currentIndex, false).EnsureCompleted();
+    }
+
+    /// <inheritdoc />
+    public override ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        return ProcessAsync(message, pipeline, currentIndex, true);
+    }
+
+    private async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        if (message.Request.Uri!.Scheme != Uri.UriSchemeHttps)
+        {
+            throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS protected (https) endpoints.");
+        }
+        Token token;
+        if (message.TryGetProperty(typeof(IScopedFlowContext), out var rawContext) && rawContext is IScopedFlowContext scopesContext)
+        {
+            var context = _flowContext.CloneWithAdditionalScopes(scopesContext.Scopes);
+            token = async ? await _tokenProvider.GetAccessTokenAsync(context, message.CancellationToken).ConfigureAwait(false) :
+            _tokenProvider.GetAccessToken(context, message.CancellationToken);
+        }
+        else
+        {
+            token = _tokenProvider.GetAccessToken(_flowContext, message.CancellationToken);
+        }
+        message.Request.Headers.Set("Authorization", $"Bearer {token.TokenValue}");
+
+        if (async)
+        {
+            await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+        }
+        else
+        {
+            ProcessNext(message, pipeline, currentIndex);
+        }
+    }
+
+    internal static IScopedFlowContext GetContext(IEnumerable<IReadOnlyDictionary<string, object>> contexts, ITokenProvider tokenProvider)
+    {
+        var type = tokenProvider.GetType();
+        // This assumes that a credential provider will only implement one flow type.
+        var credentialFlowType = type switch
+        {
+            var t when typeof(TokenProvider<IClientCredentialsFlowContext>).IsAssignableFrom(t) => typeof(IClientCredentialsFlowContext),
+            var t when typeof(TokenProvider<IAuthorizationCodeFlowContext>).IsAssignableFrom(t) => typeof(IAuthorizationCodeFlowContext),
+            var t when typeof(TokenProvider<IPasswordFlowContext>).IsAssignableFrom(t) => typeof(IPasswordFlowContext),
+            var t when typeof(TokenProvider<IImplicitFlowContext>).IsAssignableFrom(t) => typeof(IImplicitFlowContext),
+            _ => throw new InvalidOperationException("Supplied credential does not implement any supported auth flow.")
```

Comment:
Do we want to error here?  Does this matter to us if a later context works?

Reply:
I think we can remove the validation.